File uploads and PII

bdorfman-stripe opened the pull request:
Following on from https://github.com/stripe/stripe-ios/pull/516

Probably want to add client side file upload size limit detection (and maybe automatic resizing) before shipping.

bdorfman-stripe commented on Stripe/STPAPIClient.m:
```diff
@@ -260,9 +260,42 @@ - (void)createTokenWithPersonalIDNumber:(NSString *)pii completion:(__nullable S
 
 @implementation STPAPIClient (Upload)
 
+- (NSData *)dataForUploadedImage:(UIImage *)image
```
@jack-stripe Do you think this is a reasonable auto image scaling algorithm? (just keep redoing the image at 0.8x the previous scale until its small enough). Also not sure how to test this, I guess I could just check in some really large photos into our repo?

jack-stripe replied:
@bdorfman-stripe - any luck with doing this heuristically? (Based on max JPEG file sizes for certain resolutions?) Also, to test this, you could pass through to an internal `dataForUploadedImage:maxFileSize:` method, and then test that with a couple small test images.

bdorfman-stripe replied:
JPEG file size is largely dependent on how complicated the image you are encoding is, which is hard to quantify exactly.

bdorfman-stripe commented:
@bg-stripe Can you do a quick final review of this? I added a few extra tests and fixed master merge conflicts. Want to get this merged in for our next release.

bg-stripe commented on Tests/Tests/UIImage+StripeTests.m:
```diff
@@ -0,0 +1,42 @@
+
+    UIImage *testImage = [STPImageLibrary safeImageNamed:@"stp_shipping_form@3x.png"
+                                     templateIfAvailable:NO];
+
```
to make sure we're exercising the resize logic, we should probably start with an assert on the original size:
```
NSData *data = UIImageJPEGRepresentation(testImage, 0.5);
XCTAssertTrue(data.length > 50000);
```

bdorfman-stripe replied:
Ah good idea, I'll add that.

bg-stripe commented on Tests/Tests/UIImage+StripeTests.m:
```diff
@@ -0,0 +1,42 @@
+
+- (void)testJpegImageResizing {
+    // Strategy is to grab an image from our bundle and resize it to
+    // both smaller and bigger size than it already is and make sure we get
```
nit: these tests don't resize to a bigger size

bdorfman-stripe replied:
Right I meant to say putting a maxBytes larger than the initial size, not resize bigger. I will reword this comment.

bdorfman-stripe commented:
@bg-stripe Updated

bg-stripe commented on Stripe/UIImage+Stripe.m:
```diff
@@ -13,17 +13,32 @@ @implementation UIImage (Stripe)
+        // Shrink to a little bit less than we need to try to ensure we're under
+        // (otherwise its likely our first pass will be over the limit due to
+        // compression variance and floating point rounding)
+        scale = scale * sqrt(percentSmallerNeeded - (percentSmallerNeeded * 0.05));
```
ooc, what's the heuristic here?

bdorfman-stripe replied:
It's just area scaling plus a little extra. Eg if you have something that is 100x100 (area == 10k) and you want to make it area == 5k, (make it 50% of the original size) you have to scale down the height and width by sqrt(0.5) ~= 0.71 (100*0.71 ~= 71, 71*71 == 5041). 

To try to ensure it actually goes under the max byte limit after rounding and re-jpeging (since the jpeg algo isn't actually 100% linear on area) I just scaled it down a little extra than needed (and extra 5% * whatever we were scaling it by). Not sure that's the best way to do "a little extra", could try some other things if you have suggestions.

bg-stripe replied:
cool, that makes sense! was just wondering if the 5% came from anywhere in particular

bdorfman-stripe replied:
It came from me testing a few values and being like "eh this is probably fine  ¯\_(ツ)_/¯" :)
